dialer.go

```diff
@@ -68,175 +88,178 @@ func NewDialer(opts *DialerOpts) Dialer {
+		// down, as this will then spin endlessly.
+		time.Sleep(sc.sleepOnError)
+		d.pendingSessions <- sc
+	} else {
+		d.log.Debugf("Created session in %v to %#v", time.Since(start), sc)
+		d.recycleSession(s)
 	}
-	c := s.CreateStream()
-	d.returnSession(s)
-	return c, nil
 }
 
-func (d *dialer) getNumLivePending() int {
-	d.muNumLivePending.Lock()
-	numLivePending := d.numLive + d.numPending
-	d.muNumLivePending.Unlock()
-	return numLivePending
+func (d *dialer) recycleSession(s sessionIntf) {
+	if !s.allowNewStream(d.maxStreamsPerConn) {
+		d.log.Debugf("Maximum streams reached for session to %v", d.name)
+		// The default number of streams per session is 65535, so this is unlikely to be reached. If it
+		// is, we create an additional session with the same configuration as the full session.
+		d.pendingSessions <- s.getSessionConfig()
+		return
+	}
+	// We now have a new or established TCP connection/session. At this point two things can happen:
+	// 1) The connection can be closed for any reason, in which case we want to request a new one
+	// 2) A dialer can request the session. In that case, it will retrieve the session.
+	//
+	// Selecting on those two cases avoids closed sessions building up in the live session channel.
+	select {
+	case <-s.getCloseCh():
+		d.log.Debugf("Session closed before requested. Requesting new session: %#v", s.getSessionConfig())
+		d.pendingSessions <- s.getSessionConfig()
+	case d.liveSessions <- s:
+	}
 }
 
-func (d *dialer) getOrCreateSession(ctx context.Context, dial DialFN) (sessionIntf, error) {
-	newSession := func(cap int) {
-		d.muNumLivePending.Lock()
-		if d.numLive+d.numPending >= cap {
-			d.muNumLivePending.Unlock()
-			return
-		}
-		d.numPending++
-		d.muNumLivePending.Unlock()
-		go func() {
-			s, err := d.startSession(dial)
-			d.muNumLivePending.Lock()
-			d.numPending--
-			if err != nil {
-				d.muNumLivePending.Unlock()
-				return
-			}
-			d.numLive++
-			d.muNumLivePending.Unlock()
-			d.liveSessions <- s
-		}()
+func (d *dialer) Dial() (net.Conn, error) {
+	return d.DialContext(context.Background())
+}
+
+func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
+	s, err := d.getSession(ctx)
+	if err != nil {
+		return nil, err
 	}
+	go d.recycleSession(s)
+	return s.createStream(ctx), nil
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	start := time.Now()
 	for {
 		select {
 		case s := <-d.liveSessions:
-			if s.AllowNewStream(d.maxStreamsPerConn, d.idleInterval) {
-				return s, nil
+			if s.isClosed() {
+				// Closed sessions will trigger the creation of new ones, so keep waiting for a new session.
+				d.log.Debugf("Found closed session. Continuing")
+				continue
 			}
-			d.muNumLivePending.Lock()
-			d.numLive--
-			d.muNumLivePending.Unlock()
-			s.MarkDefunct()
-			newSession(minLiveConns)
-		case <-time.After(d.redialSessionInterval):
-			newSession(d.maxLiveConns)
+			d.log.Debugf("Returning session in %v", time.Since(start))
```

Comment:
Logging for each dial attempt seems excessive. Maybe change to `Tracef`?

Reply:
Yeah sorry that wasn't meant to make it into the final version - removing!